File: dijkstra/python/11779.py
from heapq import heappush,heappop
import sys
input = sys.stdin.readline

n = int(input())
m = int(input())
data = [[] for _ in range(n+1)]
for _ in range(m):
    a,b,c = map(int,input().split())
    data[a].append([c,b])
start,end = map(int,input().split())

def search(data,start,end):
    heap = []
    visited = [False for _ in range(n+1)]
    heappush(heap,[0,start,[]])
    cost_list = [0 for _ in range(n+1)]
    while heap:
        cost,to_idx,route = heappop(heap)
        if visited[to_idx]:
            continue
        visited[to_idx] = True
        if to_idx == end:
            print(cost)
            print(len(route)+1)
            print(" ".join(route),to_idx)
            return
        for next_idx in data[to_idx]:
            if visited[next_idx[1]]:
                continue
            heappush(heap,[next_idx[0]+cost,next_idx[1],route+[str(to_idx)]])
search(data,start,end)

File: dijkstra/javascript/11779.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");

const n = Number(input[0]);

const [start, end] = input.pop().split(" ").map(Number);
const arr = input.slice(2).map((row) => row.split(" ").map(Number));

const getMinDistanceNode = (arr, visited) => {
  const minNode = arr.reduce(
    (minNode, { cost }, idx) => {
      if (visited[idx]) {
        return minNode;
      }
      if (cost < minNode.cost) {
        return { idx, cost };
      }
      return minNode;
    },
    { idx: -1, cost: Infinity }
  ).idx;

  return minNode;
};

const solution = (n, arr, start, end) => {
  const route = Array.from(Array(n + 1), () => []);
  const costArr = Array.from(Array(n + 1), () => {
    return { cost: Infinity, course: [] };
  });
  costArr[start].cost = 0;

  const visited = new Array(n + 1).fill(false);

  arr.forEach(([start, end, cost]) => route[start].push([end, cost]));

  while (start !== end) {
    const { cost: currentCost, course } = costArr[start];
    visited[start] = true;
    route[start].forEach(([end, cost]) => {
      if (visited[end]) {
        return;
      }
      if (cost + currentCost < costArr[end].cost) {
        costArr[end] = { cost: cost + currentCost, course: [...course, start] };
      }
    });
    start = getMinDistanceNode(costArr, visited);
  }

  return costArr[end];
};

const { cost, course } = solution(n, arr, start, end);

console.log(cost);
console.log(course.length + 1);
console.log(course.join(" "), end);
